prow/config/config.go

```diff
@@ -913,27 +913,32 @@ func setPeriodicDecorationDefaults(c *Config, ps *Periodic) {
 // finalizeJobConfig mutates and fixes entries for jobspecs
 func (c *Config) finalizeJobConfig() error {
 	if c.decorationRequested() {
-		if c.Plank.DefaultDecorationConfig == nil {
-			return errors.New("no default decoration config provided for plank")
-		}
-		if c.Plank.DefaultDecorationConfig.UtilityImages == nil {
-			return errors.New("no default decoration image pull specs provided for plank")
-		}
-		if c.Plank.DefaultDecorationConfig.GCSConfiguration == nil {
-			return errors.New("no default GCS decoration config provided for plank")
-		}
-		if c.Plank.DefaultDecorationConfig.GCSCredentialsSecret == "" {
-			return errors.New("no default GCS credentials secret provided for plank")
-		}
+		if len(c.Plank.DefaultDecorationConfigs) == 0 {
+			if c.Plank.DefaultDecorationConfig == nil {
+				return errors.New("no default decoration config provided for plank")
+			}
 
-		if c.Plank.DefaultDecorationConfig != nil && len(c.Plank.DefaultDecorationConfigs) == 0 {
+			if err := c.Plank.DefaultDecorationConfig.Validate(); err != nil {
+				return fmt.Errorf("decoration config validation error: %v", err)
+			}
 			logrus.Warning("default_decoration_config will be deprecated on April 2020, and it will be replaced with default_decoration_configs['*'].")
-		}
 
-		if len(c.Plank.DefaultDecorationConfigs) == 0 {
 			c.Plank.DefaultDecorationConfigs = make(map[string]*prowapi.DecorationConfig)
+			c.Plank.DefaultDecorationConfigs["*"] = c.Plank.DefaultDecorationConfig
+		} else {
+			if _, ok := c.Plank.DefaultDecorationConfigs["*"]; !ok {
+				if c.Plank.DefaultDecorationConfig == nil {
+					return errors.New("default_decoration_configs specified but key '*' is missing.")
+				}
+
+				logrus.Warning("default_decoration_configs['*'] is missing, using the value from default_decoration_config instead.")
+				c.Plank.DefaultDecorationConfigs["*"] = c.Plank.DefaultDecorationConfig
+			}
+
+			if err := c.Plank.DefaultDecorationConfigs["*"].Validate(); err != nil {
```

Comment:
What is the reason we do not validate all `DefaultDecorationConfigs` ?

Reply:
I followed the same logic of merging the prowjob's decorationConfig with the defaults. That's why we need to validate only the defaults.

Reply:
Provided we validate the `"*"` defaults, is it possible for the result of the merge with a org or repo defaults to become invalid? If yes, then we can (and should) produce the merges and validate them. If not, it's okay to just validate the `"*"`.

Reply:
We are already merging the `org` or `repo` dc with the defaults in `*`.  Also after merging, the dc shouldn't become invalid. Therefore I believe that just validating `*` is enough.